libicyque.c

```diff
@@ -963,9 +992,96 @@ icq_got_buddy_icon(IcyQueAccount *ia, JsonObject *obj, gpointer user_data)
+	json_object_set_string_member(getHistoryParams, "patchVersion", "1");
+	json_object_set_string_member(getHistoryParams, "sn", chatId);
+	json_object_set_string_member(getHistoryParams, "fromMsgId", fromMsg);
+	json_object_set_int_member(getHistoryParams, "count", count);
+	
+	JsonObject *getHistoryRequest = icq_generate_robusto_request(ia, "getHistory", getHistoryParams);
+	const gchar* getHistoryRequestStr = json_object_to_string(getHistoryRequest);
+	json_object_unref(getHistoryRequest);
+	
+	icq_fetch_url_with_method(ia, "POST", ICQ_RAPI_SERVER, getHistoryRequestStr, callback, user_data);
+}
+
 static GList *valid_icyque_accounts = NULL;
 #define ICYQUE_ACCOUNT_IS_VALID(ia) (g_list_find(valid_icyque_accounts, (ia)) != NULL)
 
+static void
+icq_mark_message_as_read(IcyQueAccount *ia, const gchar *sn, const gchar *messageId)
+{
+	JsonObject *setDlgStateParams = json_object_new();
+	JsonArray *exclude = json_array_new();
+	json_object_set_string_member(setDlgStateParams, "aimSid", ia->aimsid);
+	json_object_set_array_member(setDlgStateParams, "exclude", exclude);
+	json_object_set_string_member(setDlgStateParams, "lastRead", messageId);
+	json_object_set_string_member(setDlgStateParams, "sn", sn);
+	
+	JsonObject *setDlgStateRequest = icq_generate_robusto_request(ia, "setDlgState", setDlgStateParams);
+	const gchar* setDlgStateRequestStr = json_object_to_string(setDlgStateRequest);
+	json_object_unref(setDlgStateRequest);
+	
+	icq_fetch_url_with_method(ia, "POST", ICQ_RAPI_SERVER, setDlgStateRequestStr, NULL, NULL);
+}
+
+static void
+icq_unread_message_load_cb(IcyQueAccount *ia, JsonObject *data, gpointer user_data)
+{
+	JsonObject *status = json_object_get_object_member(data, "status");
+	if(status && json_object_get_int_member(status, "code") == 20000) {
+	JsonObject *results = json_object_get_object_member(data, "results");
+		// Acquire persons first
+		JsonArray *persons = json_object_get_array_member(results, "persons");
+		const gchar* sn = NULL; // TODO: Support group chats here
+		gint i, len = json_array_get_length(persons);
+		if(len == 0) return;
+		if(len > 1) {
+			purple_connection_error(ia->pc, PURPLE_CONNECTION_ERROR_OTHER_ERROR, "Group-Conversations not yet supported. Please post the icyque input of the debug window in a github issue.");
+			return;
+		}
+		JsonObject* firstPerson = json_array_get_object_element(persons, 0);
+		sn = json_object_get_string_member(firstPerson, "sn");
+	
+		JsonArray *messages = json_object_get_array_member(results, "messages");
+		len = json_array_get_length(messages);
+	
+		for (i = (len - 1); i >= 0; i--) {
+			JsonObject *message = json_array_get_object_element(messages, i);
+			gint64 time = json_object_get_int_member(message, "time");
+			const gchar* text = json_object_get_string_member(message, "text");
+			gchar *escaped_text = purple_markup_escape_text(text, -1);
+			purple_serv_got_im(ia->pc, sn, escaped_text, PURPLE_MESSAGE_RECV, (time_t) time);

```

Comment:
Does this api also display self-sent messages?  If so, you'll need to use similar code to elsewhere where `purple_serv_got_im` is used to display self-sent messages differently to messages sent by someone else

Reply:
No, it can't. When you send a message, every received message in the history up to the point where your message was received by the server is marked as unread. So it should be impossible that this request contains your own messages.
There might be a race-condition though. Following scenario:
```
purple: [session start] -> [fetchEvent] ->	                                          [request unread messages]
client1:                                 [send message in conversation with unread messages]
```
Since I am using "-1" as `fromMessageId`, the last message that the server knows of is returned. In the scenario above, this would be the message sent by client1, instead of the last received unread message.
I will add a `TODO` for this, for now.

Reply:
What about when you send a message from a different device like the webclient or a phone and then this history fetching displays it in Pidgin?